csi/node_server.go

```diff
@@ -60,7 +60,7 @@ func (ns *NodeServer) NodePublishVolume(ctx context.Context, req *csi.NodePublis
 
 	// Check volume frontend settings
 	if existVol.DisableFrontend || existVol.Frontend != string(types.VolumeFrontendBlockDev) {
-		return nil, status.Errorf(codes.InvalidArgument, "There is no block device frontend for volume %s", req.GetVolumeId())
+		return nil, status.Errorf(codes.FailedPrecondition, "There is no block device frontend for volume %s", req.GetVolumeId())
```

Comment:
I had a quick look at the csi spec to see if they deal with this case, since the CO needs to ensure NodePublishVolume call is only made after successfully calling ControllerPublishVolume.
https://github.com/container-storage-interface/spec/blob/master/spec.md#nodepublishvolume-errors

They don't so it's okay for now to return `InvalidArgument` for both scenarios, since we now guarantee that we have a valid endpoint before returning sucess in the ControllerPublishVolume call.

Reply:
Thanks a lot for the detailed explanation. I agree with you that we can use `InvalidArgument`, for line 68, how about `Unavailable`? Like you said in ControllerPublishVolume, we have guaranteed we will have a valid endpoint, if we see any empty endpoint here at line 68, does that mean unavailable? Thanks.

Reply:
That's a good thought, `Unavailable` is a transient error state. If we assume that the volume will eventually become attached and has a valid controller then returning `Unavailable` is fine.
https://github.com/grpc/grpc-go/blob/master/codes/codes.go#L105

I have no strong feelings about this one, but I feel like `InvalidArgument` might be better, because the system should already be in this state (attached & endpoint available) after returning from Controller.PublishVolume and if it's not then it will not get into that state either. So `Unavailable` would signal hey retry this call eventually it will work.